runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -0,0 +1,674 @@
+     */
+    @Accessor
+    public List<V> scanAndFilter(Predicate<? super V> p) {
+        return mainMap.values().parallelStream().filter(p).collect(Collectors.toList());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public Collection<Map.Entry<K, V>> scanAndFilterByEntry(Predicate<? super Map.Entry<K, V>>
+                                                                    entryPredicate) {
+        return mainMap.entrySet().parallelStream().filter(entryPredicate).collect(Collectors
+                .toCollection(ArrayList::new));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @MutatorAccessor(name = "remove", undoFunction = "undoRemove",
+                                undoRecordFunction = "undoRemoveRecord")
+    @SuppressWarnings("unchecked")
+    public V remove(@ConflictParameter Object key) {
+        V previous =  mainMap.remove(key);
+        unmapSecondaryIndexes((K) key, previous);
+        return previous;
+    }
+
+    @DontInstrument
+    protected V undoRemoveRecord(CorfuTable<K, V, F, I> table, K key) {
+        return table.mainMap.get(key);
+    }
+
+    @DontInstrument
+    protected void undoRemove(CorfuTable<K, V, F, I> table, V undoRecord, K key) {
+        if (undoRecord == null) {
+            V previous =  table.mainMap.remove(key);
+            table.unmapSecondaryIndexes(key, previous);
+        } else {
+            V previous = table.mainMap.put(key, undoRecord);
+            if (!table.indexFunctions.isEmpty()) {
+                table.unmapSecondaryIndexes(key, previous);
+                table.mapSecondaryIndexes(key, undoRecord);
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Mutator(name = "remove", noUpcall = true)
+    public void delete(@ConflictParameter K key) {
+        V previous =  mainMap.remove(key);
+        unmapSecondaryIndexes(key, previous);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Mutator(name = "putAll",
+            undoFunction = "undoPutAll",
+            undoRecordFunction = "undoPutAllRecord",
+            conflictParameterFunction = "putAllConflictFunction")
+    public void putAll(@Nonnull Map<? extends K, ? extends V> m) {
+        // If we have no index functions, then just directly put all
+        if (indexFunctions.isEmpty()) {
+            mainMap.putAll(m);
+        } else {
+            // Otherwise we must update all secondary indexes
+            // TODO: Do this in parallel (need to acquire update locks, potentially)
+            m.entrySet().stream()
+                    .forEach(e -> {
+                        V previous = mainMap.put(e.getKey(), e.getValue());
+                        unmapSecondaryIndexes(e.getKey(), previous);
+                        mapSecondaryIndexes(e.getKey(), e.getValue());
+                    });
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Mutator(name = "clear", reset = true)
+    public void clear() {
+        mainMap.clear();
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this lambda with a method reference. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)